### PR TITLE
feat: remove node dependencies from respect-core

### PR DESCRIPTION
**Pull request by DmitryAnansky**
## What/Why/How?

Remove node dependencies from respect-core:

- [x] generate-arazzo
- [x] respect

## Check yourself

- [ ] Code changed? - Tested with redoc (internal)
- [x] All new/updated code is covered with tests
- [ ] New package installed? - Tested in different environments (browser/node)

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/respect-core/src/modules/__tests__/flow-runner/context/create-test-context.test.ts [hunk 1]
```diff
@@ -47,14 +48,15 @@ describe('createTestContext', () => {
     } as unknown as TestDescription;
 
     const options = {
-      workflowPath: 'test.test.yaml',
+      workflowPath: 'modules/description-parser/test.test.yaml',
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/__tests__/flow-runner/context/create-test-context.test.ts]: External resolver seems to have different behavior when running from test file as a code.
Should be investigated. Does not affect e2e.

File: packages/respect-core/src/handlers/run.ts [hunk 2]
```diff
@@ -23,124 +11,70 @@ export type RespectOptions = {
...
-    const { files } = argv;
-    const runAllFilesResult = [];
-
-    if (files.length > 1 && harOutputFile) {
-      // TODO: implement multiple run files HAR output
-      throw new Error(
-        'Currently only a single file can be run with --har-output. Please run a single file at a time.'
-      );
-    }
+export async function handleRun(options: RespectOptions): Promise<RunFileResult[]> {
```
Comment (DmitryAnansky) [inline packages/respect-core/src/handlers/run.ts]: TODO: find a better name for `handleRun`.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -272,6 +272,13 @@ Prepare the local build, go to the `benchmark` folder, clean it up, do the prepa
 
 You might need to adjust the CLI versions that need to be tested in the `benchmark/package.json` file.
 
+### Manual tests
+
+What should be verified when changes applied to the `respect-core` package:
```
Comment (JLekawa) [inline CONTRIBUTING.md:277]:
```suggestion
What needs to be verified when changes are applied to the `respect-core` package:
```

**Review by JLekawa — APPROVED** [review]

File: packages/core/src/logger.ts [hunk 4]
```diff
@@ -2,6 +2,8 @@ import * as colorette from 'colorette';
 import { isBrowser } from './env.js';
 import { identity } from './utils.js';
 
+export const RESET_ESCAPE_CODE_IN_TERMINAL = process.env.NO_COLOR ? '' : '\x1B[0m';
```
Comment (DmitryAnansky) [inline packages/core/src/logger.ts]: add isBrowser() check for process

File: packages/cli/src/commands/respect/index.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (options.files.length > 1 && argv['har-output']) {
+      // TODO: implement multiple run files HAR output
+      throw new Error(
+        'Currently only a single file can be run with --har-output. Please run a single file at a time.'
+      );
+    }
+
+    const startedAt = performance.now();
+    // TODO: continue refactoring
```
Comment (DmitryAnansky) [inline packages/cli/src/commands/respect/index.ts]: remove todo

File: packages/respect-core/package.json [hunk 6]
```diff
@@ -31,35 +31,23 @@
...
-    "@types/har-format": "^1.2.16",
-    "@types/js-yaml": "4.0.9",
     "@types/json-pointer": "1.0.34",
-    "@types/set-cookie-parser": "2.4.10",
     "json-schema-to-ts": "^3.0.1",
     "typescript": "5.6.2"
   },
   "dependencies": {
     "@faker-js/faker": "^7.6.0",
+    "@noble/hashes": "^1.8.0",
```
Comment (DmitryAnansky) [inline packages/respect-core/package.json:40]: Add to use platform independent md5 and sh256

File: packages/respect-core/src/modules/cli-output/display-checks.ts [hunk 7]
```diff
@@ -3,37 +3,45 @@ import { outdent } from 'outdent';
...
-const logger = DefaultLogger.getInstance();
 const MAX_CRITERIA_CONDITION_DISPLAY_LENGTH = 50;
 
-export function displayChecks(
-  testNameToDisplay: string,
-  checks: Check[],
-  verboseLogs?: VerboseLog,
-  verboseResponseLogs?: VerboseLog
-) {
+export function displayChecks({
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/cli-output/display-checks.ts:12]: Rename parent folder to `logger-output`

File: packages/respect-core/src/handlers/generate.ts [hunk 8]
```diff
@@ -1,36 +1,16 @@
...
-import { writeFileSync } from 'node:fs';
-import { stringifyYaml } from '../utils/yaml.js';
 import { generateArazzoDescription } from '../modules/arazzo-description-generator/index.js';
-import { DefaultLogger } from '../utils/logger/logger.js';
-import { exitWithError } from '../utils/exit-with-error.js';
-import { type CommandArgs } from '../types.js';
+import { type BaseResolver, type CollectFn, type Config } from '@redocly/openapi-core';
 
-export type GenerateArazzoFileOptions = {
+export type GenerateArazzoOptions = {
```
Comment (DmitryAnansky) [inline packages/respect-core/src/handlers/generate.ts]: Move to top level and delete `handlers` folder

File: packages/respect-core/src/modules/context-parser/parse-request-body.ts [hunk 9]
```diff
@@ -18,49 +10,42 @@ export function stripFileDecorator(payload: string) {
...
-    access(filePath, constants.F_OK | constants.R_OK, (err) => {
-      if (err) {
-        reject(new Error(`File ${filePath} doesn't exist or isn't readable.`));
-      } else {
-        formData.append(key, createReadStream(filePath));
-        resolve();
-      }
-    });
-  });
+  formData.append(key, await ctx.requestFileLoader.getFileBody(filePath));
```
Comment (DmitryAnansky) [inline packages/respect-core/src/modules/context-parser/parse-request-body.ts]: check if should be used directrly from options

File: packages/respect-core/src/utils/cli-outputs.ts [hunk 10]
```diff
@@ -30,58 +27,70 @@ export function indent(str: string, level: number) {
...
   logger.printNewLine();
   logger.printNewLine();
 }
 
-export function printWorkflowSeparator(
-  fileName: string,
-  workflowName: string | undefined,
-  skipLineSeparator?: boolean
-) {
+export function printWorkflowSeparator({
```
Comment (DmitryAnansky) [inline packages/respect-core/src/utils/cli-outputs.ts:36]: Move to other logger outputs

File: packages/cli/src/__tests__/commands/generate-arazzo.test.ts [hunk 11]
```diff
@@ -0,0 +1,104 @@
+import { GenerateArazzoCommandArgv, handleGenerateArazzo } from '../../commands/generate-arazzo.js';
+import { generate } from '@redocly/respect-core';
+import { vi, describe, it, expect, beforeEach } from 'vitest';
+import { Config } from '@redocly/openapi-core';
+
+// Mock the generate function
```
Comment (tatomyr) [inline packages/cli/src/__tests__/commands/generate-arazzo.test.ts]: Please remove the comments. The code should be self-descriptive.

File: packages/cli/src/__tests__/commands/respect/har-logs/create-har-log.test.ts [hunk 12]
```diff
@@ -1,11 +1,11 @@
-import { createHarLog } from '../../har-logs/index.js';
+import { createHarLog } from '../../../../commands/respect/har-logs/har-logs.js';
 
 describe('createHarLog', () => {
   it('should create a har log', () => {
     const mockDate = new Date('2024-01-01T00:00:00.000Z');
     vi.spyOn(global, 'Date').mockImplementation(() => mockDate as any);
 
-    const harLog = createHarLog();
+    const harLog = createHarLog({ version: '1.0.0' });
```
Comment (tatomyr) [inline packages/cli/src/__tests__/commands/respect/har-logs/create-har-log.test.ts:8]: What is the version?

File: packages/cli/src/__tests__/commands/respect/respect.test.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    constants: {
+      F_OK: 0,
+      R_OK: 4,
+    },
+  };
+});
+
+// Mock the run function
+vi.mock('@redocly/respect-core', async () => {
+  const actual = await vi.importActual('@redocly/respect-core');
```
Comment (tatomyr) [inline packages/cli/src/__tests__/commands/respect/respect.test.ts]: Does it infer the types? If not, you can use a generic like this for better compatibility:

```suggestion
  const actual = await vi.importActual<typeof import('@redocly/respect-core')>('@redocly/respect-core');
```

File: packages/cli/src/commands/respect/har-logs/with-har.ts [hunk 14]
```diff
@@ -19,11 +18,14 @@ import { buildResponseCookies } from './helpers/build-response-cookies.js';
...
   harEntryMap?: Map<string, any>;
 }
 
-export const withHar: WithHar = function (baseFetch: any, defaults: any = {}): any {
+export const withHar: WithHar = function <T extends typeof fetch>(
+  baseFetch: any,
+  defaults: any = {}
+): T {
   withHar.harEntryMap = harEntryMap;
   return async function fetch(input: any, options: any = {}): Promise<any> {
```
Comment (tatomyr) [inline packages/cli/src/commands/respect/har-logs/with-har.ts:30]:
```suggestion
  return async function fetch(input: any, options: any = {}): Promise<T> {
```

Will this work instead of https://github.com/Redocly/redocly-cli/pull/2159/files#diff-6f2c7638aa3216cf7e9a99e92593236ed73cd8bc5b77b5ffd1f6d81ce6688116R235?

File: packages/cli/src/index.ts [hunk 15]
```diff
@@ -668,7 +672,8 @@ yargs(hideBin(process.argv))
           },
         }),
     (argv) => {
-      commandWrapper(handleEject)(argv as Arguments<EjectOptions>);
+      process.env.REDOCLY_CLI_COMMAND = 'eject';
```
Comment (tatomyr) [inline packages/cli/src/index.ts]:
```suggestion
```

Is there a reason in adding those back?

File: packages/respect-core/src/index.ts [hunk 16]
```diff
@@ -1,4 +1,8 @@
-export { handleGenerate, handleRun } from './handlers/index.js';
-export type { GenerateArazzoFileOptions } from './handlers/generate.js';
-export type { RespectOptions } from './handlers/run.js';
-export type { JsonLogs } from './types.js';
+export { generate } from './generate.js';
+export { run } from './run.js';
+export type { GenerateArazzoOptions } from './generate.js';
+export type { RespectOptions } from './run.js';
```
Comment (tatomyr) [inline packages/respect-core/src/index.ts]: I'd suggest sticking to this pattern to avoid duplicated imports (for me this looks nicer):

```suggestion
export { generate, type GenerateArazzoOptions } from './generate.js';
export { run, type RespectOptions } from './run.js';
```

What do you think?

File: packages/respect-core/src/modules/__tests__/config-parser/handle-request-body-replacements.test.ts [hunk 17]
```diff
@@ -45,26 +46,26 @@ describe('handlePayloadReplacements', () => {
...
+      handlePayloadReplacements({ payload, replacements, expressionContext, logger })
+    ).toThrowError('Invalid JSON Pointer: /bar');
   });
 
   it('should throw an error if the target is not a string', () => {
     const payload = {
       foo: 'bar',
     };
-    const replacements = [{ target: 1, value: 'baz' }];
+    const replacements = [{ target: 1, value: 'baz' }] as any;
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/config-parser/handle-request-body-replacements.test.ts:58]: What was the issue?

File: packages/respect-core/src/modules/__tests__/config-parser/resolve-reusable-component-item.test.ts [hunk 18]
```diff
@@ -1,11 +1,16 @@
...
 
 describe('resolveReusableComponentItem', () => {
   it('should return parameter if not reference', () => {
     expect(
-      resolveReusableComponentItem({ in: 'query', name: 'test', value: 'test' }, {} as any)
+      resolveReusableComponentItem({ in: 'query', name: 'test', value: 'test' }, {
+        options: {
+          logger,
+        },
+      } as unknown as TestContext)
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/config-parser/resolve-reusable-component-item.test.ts]: ?

```suggestion
      } as Partial<TestContext> as TestContext)
```

Or what is the issue?

File: packages/respect-core/src/modules/context-parser/parse-request-body.ts [hunk 19]
```diff
@@ -122,26 +110,26 @@ export async function parseRequestBody(
...
     const workflowFilePath = path.resolve(ctx.options.workflowPath);
 
-    await getRequestBodyMultipartFormData(payload, formData, workflowFilePath);
+    await getRequestBodyMultipartFormData(payload, formData, workflowFilePath, ctx);
 
     return {
       ...stepRequestBody,
       payload: formData,
-      contentType: `multipart/form-data; boundary=${formData.getBoundary()}`,
+      // contentType: `multipart/form-data; boundary=${formData.getBoundary()}`,
```
Comment (tatomyr) [inline packages/respect-core/src/modules/context-parser/parse-request-body.ts]: Why it's commented?

File: packages/respect-core/src/modules/context-parser/parse-request-body.ts [hunk 20]
```diff
@@ -70,30 +55,25 @@ const appendObjectToFormData = (
...
+    appendObjectToFormData(promises, formData, payload, workflowFilePath, ctx);
     await Promise.all(promises);
   }
 };
 
-const getRequestBodyOctetStream = async (payload: RequestBody['payload']) => {
+const getRequestBodyOctetStream = async (payload: RequestBody['payload'], ctx: TestContext) => {
   if (typeof payload === 'string' && payload.startsWith('$file(') && payload.endsWith(')')) {
-    const filePath = path.resolve(__internalDirname, '../', stripFileDecorator(payload));
+    // fixme, remove this
```
Comment (tatomyr) [inline packages/respect-core/src/modules/context-parser/parse-request-body.ts]: Please address the fixme.

File: packages/respect-core/src/modules/context-parser/get-value-from-context.ts [hunk 21]
```diff
@@ -190,30 +219,39 @@ export function getFakeData(pointer: string, ctx: TestContext | RuntimeExpressio
...
 
-function modifyJSON(value: any, ctx: TestContext | RuntimeExpressionContext): any {
+function modifyJSON(
+  value: any,
+  ctx: TestContext | RuntimeExpressionContext,
+  logger: LoggerInterface
+): any {
   if (typeof value === 'string') {
-    if (value) return getValueFromContext(value, ctx);
+    if (value) return getValueFromContext({ value, ctx, logger });
```
Comment (tatomyr) [inline packages/respect-core/src/modules/context-parser/get-value-from-context.ts:228]: Have you considered making the logger a part of ctx?

File: packages/respect-core/src/modules/__tests__/flow-runner/call-api-and-analyze-results.test.ts [hunk 22]
```diff
@@ -7,12 +6,20 @@ import {
...
-vi.mock('undici');
+const originalFetch = global.fetch;
+
+beforeAll(() => {
+  // Reset fetch mock before each test
+  global.fetch = vi.fn();
+});
+
+afterAll(() => {
+  // Restore original fetch after each test
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/flow-runner/call-api-and-analyze-results.test.ts]: We already do mocks resetting (see Vite config). Also, please remove redundant comments.

Comment (tatomyr) [reply packages/respect-core/src/modules/__tests__/flow-runner/call-api-and-analyze-results.test.ts]: BTW, can we pass a mocked fetch instead of mocking the real one?

File: packages/respect-core/src/modules/__tests__/flow-runner/run-step.test.ts [hunk 23]
```diff
@@ -14,24 +15,19 @@ import {
...
 
-vi.mock('../../cli-output/index.js', async () => {
-  const actual = await vi.importActual('../../cli-output/index.js');
-
-  return {
-    ...actual,
-    displayChecks: vi.fn(),
-  };
-});
+// Add this mock for the direct import path used by helpers.ts
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/flow-runner/run-step.test.ts]: Who should add this mock?

File: packages/respect-core/src/modules/__tests__/flow-runner/runner/run-test-file.test.ts [hunk 24]
```diff
@@ -1,17 +1,14 @@
...
-  const originalModule = vi.importActual('../../../../utils/yaml.js');
-
-  return {
-    ...originalModule, // In case there are other exports you want to preserve
-    readYaml: vi.fn(),
-  };
-});
 
 vi.mock('@redocly/openapi-core', async () => {
   const originalModule = await vi.importActual('@redocly/openapi-core');
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/flow-runner/runner/run-test-file.test.ts]: Does it infer the types?

File: packages/respect-core/src/modules/__tests__/flow-runner/runner/run-test-file.test.ts [hunk 25]
```diff
@@ -604,25 +622,28 @@ describe('runTestFile', () => {
...
     );
 
-    vi.mocked(readYaml).mockResolvedValue(mockDocument.parsed);
     vi.mocked(lint).mockResolvedValueOnce([]);
     vi.mocked(bundle).mockResolvedValueOnce({
       bundle: {
         parsed: mockDocument.parsed,
       },
     } as any);
+    vi.mocked(bundle).mockResolvedValueOnce(undefined as any);
```
Comment (tatomyr) [inline packages/respect-core/src/modules/__tests__/flow-runner/runner/run-test-file.test.ts:632]: Could it return `undefined`?

**Review by tatomyr — COMMENTED** [review]
Reviewed the first batch.
Will do another round later today.

File: packages/cli/src/__tests__/commands/respect/har-logs/create-har-log.test.ts (hunk 12 above)
Comment (DmitryAnansky) [reply packages/cli/src/__tests__/commands/respect/har-logs/create-har-log.test.ts:8]: Just dummy version for test.

File: packages/cli/src/commands/respect/har-logs/with-har.ts (hunk 14 above)
Comment (DmitryAnansky) [reply packages/cli/src/commands/respect/har-logs/with-har.ts:30]: No it won't with-har is a lib fork that was designed for node-fetch lib and adopted to work with native node fetch but  it can't handle browser fetch or any other replacements. We would need to review har-logs implementation in the future.

File: packages/cli/src/index.ts (hunk 15 above)
Comment (DmitryAnansky) [reply packages/cli/src/index.ts]: My mistake, I guess added during rebase, removing ....

File: packages/respect-core/src/modules/__tests__/config-parser/handle-request-body-replacements.test.ts (hunk 17 above)
Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/config-parser/handle-request-body-replacements.test.ts:58]: I try to pass here number but actual type expect string, so I need to cast to see negative scenario handling correctly and not only by TS check.

File: packages/respect-core/src/modules/__tests__/config-parser/resolve-reusable-component-item.test.ts (hunk 18 above)
Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/config-parser/resolve-reusable-component-item.test.ts]: This is not applicable to all test cases, we don't always describe inner properties in its complete form it Test, as it will just increase text size and don't bring any value for particular test case.
I will update some places but not all.

File: packages/respect-core/src/modules/context-parser/get-value-from-context.ts (hunk 21 above)
Comment (DmitryAnansky) [reply packages/respect-core/src/modules/context-parser/get-value-from-context.ts:228]: The reason for this is the type of ctx that `getValueFromContext` can accept as parameter:
`ctx: TestContext | RuntimeExpressionContext;` and RuntimeExpressionContext does not have options.

File: packages/respect-core/src/modules/__tests__/flow-runner/runner/run-test-file.test.ts (hunk 25 above)
Comment (DmitryAnansky) [reply packages/respect-core/src/modules/__tests__/flow-runner/runner/run-test-file.test.ts:632]: For this test it does not matter, as I just need to mock any return and test another negative scenario.

File: packages/cli/src/index.ts [hunk 26]
```diff
@@ -760,8 +764,8 @@ yargs(hideBin(process.argv))
         });
     },
     async (argv) => {
-      const { handleRun } = await import('@redocly/respect-core');
-      commandWrapper(handleRun)(argv as Arguments<RespectOptions>);
+      process.env.REDOCLY_CLI_COMMAND = 'respect';
```
Comment (tatomyr) [inline packages/cli/src/index.ts]: Clean up this.

File: packages/respect-core/src/modules/flow-runner/prepare-request.ts [hunk 27]
```diff
@@ -161,6 +162,7 @@ export async function prepareRequest(
         payload: parameter.value,
         context: expressionContext,
         // contentType,
```
Comment (tatomyr) [inline packages/respect-core/src/modules/flow-runner/prepare-request.ts]: Remove this

File: packages/respect-core/src/modules/flow-runner/runner.ts [hunk 28]
```diff
@@ -1,108 +1,76 @@
...
-      clientKey: argv['client-key'],
-      caCert: argv['ca-cert'],
-    },
-    maxSteps: argv['max-steps'],
-    maxFetchTimeout: argv['max-fetch-timeout'],
-    executionTimeout: argv['execution-timeout'],
+export async function runTestFile(options: RunOptions, collectSpecData?: CollectFn) {
+  const workflowOptions = {
+    ...options,
+    workflowPath: options.file, // filePath or documentPath
```
Comment (tatomyr) [inline packages/respect-core/src/modules/flow-runner/runner.ts]:
```suggestion
    filePath: options.file, // filePath or documentPath
```

File: packages/respect-core/src/modules/flow-runner/runner.ts [hunk 29]
```diff
@@ -1,108 +1,76 @@
...
-    severity,
-    mutualTls: {
-      clientCert: argv['client-cert'],
-      clientKey: argv['client-key'],
-      caCert: argv['ca-cert'],
-    },
-    maxSteps: argv['max-steps'],
-    maxFetchTimeout: argv['max-fetch-timeout'],
-    executionTimeout: argv['execution-timeout'],
+export async function runTestFile(options: RunOptions, collectSpecData?: CollectFn) {
```
Comment (tatomyr) [inline packages/respect-core/src/modules/flow-runner/runner.ts:32]:
```suggestion
export async function runTestFile({ file: filePath, ...options} : RunOptions, collectSpecData?: CollectFn) {
```

Comment (DmitryAnansky) [reply packages/respect-core/src/modules/flow-runner/runner.ts:32]: I will leave prev version.
